pipeline/README.md

```diff
@@ -0,0 +1,135 @@
+
+## CI/CD Pipeline Diagram
+
+
+![Reference Architecture - Real-time File Processing CI/CD Pipeline](img/lambda-refarch-fileprocessing-simple-pipeline.png)
+
+
+## Pipeline Components
+
+
+### CloudFormation Template
+
+
+pipeline/pipeline.yml is a CloudFormation template that will deploy all the required pipeline components. Once the stack has deployed the Pipeline will automatically execute and deploy the Serverless Application. See getting started for information on how to deploy the template.
+
+
+#### Deployed Resources
+
+
+* Pipeline S3 bucket, used to store pipeline artefacts that are passed between stages.
+* CodePipeline
+* CodeBuild Project
+* Roles for CodePipeline, CodeBuild and the CloudFormation Deployment
+
+
+### Source
+
+
+For this application we are hosting our source code in GitHub. Other [Source Integrations](https://docs.aws.amazon.com/codepipeline/latest/userguide/integrations-action-type.html#integrations-source) are available however this template focuses on GitHub. Whenever an update is pushed to the GitHub branch being
+monitored our pipeline will begin executing. The source stage will connect to GitHub using the credentials provided and clone the branch into our pipeline artefact bucket for use in the other stages. 
+
+
+### Build
+
+
+In order to run our SAM build and SAM package commands we are using [CodeBuild](https://aws.amazon.com/codebuild/), a fully managed continuous integration service . Codebuild allows us to perform a sequence of commands that we define in the [BuildSpec.yml](https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html)
+file that will execute inside the [build environment](https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref.html) we define using a docker container. For this project we are using the Amazon Linux 2 version 1.0 container with Python 3.7.
+
+Within the buildspec.yml we are:
+
+* Updating SAM to the latest version
+* Running SAM build as per the manual deployment
+* Running SAM Package again as per the manual deployment steps
+* Instructing CodeBuild to pass the output template back to the Pipeline for use in the deployment stage. 
+
+
+
+### Deploy
+
+
+To deploy our application stack we are not using SAM Deploy, instead we are opting to use the CodePipeline native support for CloudFormation. The pipeline has a role it use with appropriate permissions to deploy the template created by the SAM package step which will create a stack containing the resources defined in our SAM Template. We are using [change sets](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-changesets.html) and [approval actions](https://docs.aws.amazon.com/codepipeline/latest/userguide/approvals-action-add.html) to demonstrate a manual approval workflow. The first deployment will not require approval however subsequent updates will.
```

Comment:
Any reason to not use SAM?  Granted, it's just an alias for the aws cloudformation commands under the covers but there is native integration between CodeDeploy and SAM. https://docs.aws.amazon.com/codedeploy/latest/userguide/tutorial-lambda-sam.html

Reply:
Where is the code to do approval actions?